int64 num
bool signal
